pkg/agent/openflow/pipeline.go

```diff
@@ -2915,35 +2923,45 @@ func (f *featureService) gatewaySNATFlows() []binding.Flow {
 	cookieID := f.cookieAllocator.Request(f.category).Raw()
 	var flows []binding.Flow
 	for _, ipProtocol := range f.ipProtocols {
-		flows = append(flows,
-			// This generates the flow to match the first packet of hairpin connection initiated through the Antrea gateway.
-			// ConnSNATCTMark and HairpinCTMark will be loaded in DNAT CT zone.
-			ServiceMarkTable.ofTable.BuildFlow(priorityNormal).
-				Cookie(cookieID).
-				MatchProtocol(ipProtocol).
-				MatchCTStateNew(true).
-				MatchCTStateTrk(true).
-				MatchRegMark(GatewayHairpinRegMark).
-				Action().CT(true, ServiceMarkTable.GetNext(), f.dnatCtZones[ipProtocol], f.ctZoneSrcField).
-				LoadToCtMark(ConnSNATCTMark).
-				LoadToCtMark(HairpinCTMark).
-				CTDone().
-				Done(),
+		// This generates the flow to match the first packet of hairpin connection initiated through the Antrea gateway.
+		// ConnSNATCTMark and HairpinCTMark will be loaded in DNAT CT zone.
+		flows = append(flows, ServiceMarkTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchCTStateNew(true).
+			MatchCTStateTrk(true).
+			MatchRegMark(FromGatewayRegMark).
+			MatchRegMark(ToGatewayRegMark).
+			Action().CT(true, ServiceMarkTable.GetNext(), f.dnatCtZones[ipProtocol], f.ctZoneSrcField).
+			LoadToCtMark(ConnSNATCTMark).
+			LoadToCtMark(HairpinCTMark).
+			CTDone().
+			Done())
+
+		var pktDstRegMarks []*binding.RegMark
+		if f.networkConfig.TrafficEncapMode.SupportsEncap() {
+			pktDstRegMarks = append(pktDstRegMarks, ToTunnelRegMark)
+		}
+		if f.networkConfig.TrafficEncapMode.SupportsNoEncap() && runtime.IsWindowsPlatform() {
+			pktDstRegMarks = append(pktDstRegMarks, ToUplinkRegMark)
+		}
+		for _, pktDstRegMark := range pktDstRegMarks {
 			// This generates the flow to match the first packet of NodePort / LoadBalancer connection initiated through the
-			// Antrea gateway and externalTrafficPolicy of the Service is Cluster. ConnSNATCTMark will be loaded in DNAT
-			// CT zone.
-			ServiceMarkTable.ofTable.BuildFlow(priorityLow).
+			// Antrea gateway and externalTrafficPolicy of the Service is Cluster, and the selected Endpoint is on remote
+			// Node, then ConnSNATCTMark will be loaded in DNAT CT zone, indicating that SNAT is required for the connection.
+			flows = append(flows, ServiceMarkTable.ofTable.BuildFlow(priorityLow).
```

Comment:
Does it still need to be low?

Reply:
Yes, some connection like gateway hairpin cases also has `FromGatewayRegMark` and `ToClusterServiceRegMark`

Reply:
But these low flows match ToXXXRegMark so they cannot match hairpin cases?
gateway hairpin case has `FromGatewayRegMark` and `ToGatewayRegMark`
gateway->service->tunnel has `FromGatewayRegMark` and `ToTunnelRegMark`
gateway->service->uplink has `FromGatewayRegMark` and `ToUplinkRegMark`

Reply:
@tnqn You are right, it doesn't need to be low since the match condition are mutually exclusive.

- FromGatewayRegMark, ToGatewayRegMark
- FromGatewayRegMark, ToTunnelRegMark
- FromGatewayRegMark, ToUplinkRegMark